grakn-graql/src/main/java/ai/grakn/graql/internal/gremlin/GreedyTraversalPlan.java

```diff
@@ -136,6 +174,123 @@ public static GraqlTraversal createTraversal(PatternAdmin pattern, EmbeddedGrakn
+                relationshipRolePlayerMap.get(relationship).forEach(
+                        rolePlayer -> {
+                            if (instanceVarTypeMap.containsKey(rolePlayer)) {
+                                numRolePlayersHaveType[0]++;
+                            }
+                        });
+                if (numRolePlayersHaveType[0] < 2) {
+                    iterator.remove();
+                }
+            }
+        }
+        if (relationshipRolePlayerMap.isEmpty()) return;
+
+        // for each type, get all possible relationship type it could be in
+        Map<Type, Set<RelationshipType>> relationshipMap = new HashMap<>();
+        getAllPossibleRelationships(relationshipMap, tx.admin().getMetaEntityType());
+        getAllPossibleRelationships(relationshipMap, tx.admin().getMetaAttributeType());
+        getAllPossibleRelationships(relationshipMap, tx.admin().getMetaRelationType());
+
+        relationshipRolePlayerMap.asMap().forEach((relationshipVar, rolePlayerVars) -> {
+            Iterator<Var> rolePlayerVarIterator = rolePlayerVars.iterator();
+
+            Var firstRolePlayer = rolePlayerVarIterator.next();
+            while (!instanceVarTypeMap.containsKey(firstRolePlayer)) {
+                firstRolePlayer = rolePlayerVarIterator.next();
+            }
+
+            Set<Type> possibleRelationShipTypes = getAllPossibleRelationShipTypesOfARolePlayer(
+                    firstRolePlayer, instanceVarTypeMap, relationshipMap);
+            while (rolePlayerVarIterator.hasNext()) {
+                Var nextRolePlayer = rolePlayerVarIterator.next();
+                if (instanceVarTypeMap.containsKey(nextRolePlayer)) {
+
+                    // compute the intersection of possible relationship types
+                    possibleRelationShipTypes.retainAll(getAllPossibleRelationShipTypesOfARolePlayer(
+                            nextRolePlayer, instanceVarTypeMap, relationshipMap));
+                }
+            }
+
+            //TODO: if possibleRelationShipTypes here is empty, the query will not match any data
+            if (possibleRelationShipTypes.size() == 1) {
+
+                Type relationshipType = possibleRelationShipTypes.iterator().next();
+                Label label = relationshipType.getLabel();
+                Var labelVar = var();
+                allFragments.add(Fragments.label(LabelProperty.of(label), labelVar, ImmutableSet.of(label)));
+                allFragments.addAll(EquivalentFragmentSets.isa(IsaProperty.of(label(label).admin()),
+                        relationshipVar, labelVar, relationshipType.isImplicit()).fragments());
+            }
+        });
+    }
+
+    private static Set<Type> getAllPossibleRelationShipTypesOfARolePlayer(
+            Var rolePlayer, Multimap<Var, Type> instanceVarTypeMap, Map<Type, Set<RelationshipType>> relationshipMap) {
+
+        Iterator<Type> typeIterator = instanceVarTypeMap.get(rolePlayer).iterator();
```

Comment:
how do we guarantee that a roleplayer has a type? Does having multiple possibilities allow for types being part of the same hierarchy?

Reply:
We do the checking before calling this method